### ISSUE TITLE
Partnership With Codewars

### ISSUE THREAD

**Issue by Philosophist** [Anuken/Mindustry-Suggestions#2051]
**Describe the content or mechanics you are proposing.**
I recommend reaching out to https://www.codewars.com/ to offer a fun set of challenges with Mindustry Logic that would give visual results and combine the best of both worlds.

**Describe how you think this content will improve the game. If you're proposing new content, mention how it may add more gameplay options or how it will fill a new niche.**
Probably only able to achieve this down the road, but I want Anuken to have this kind of option ready when the opportunity hits. Codewars offers a really functional and fun set up for code challenges for all levels. Once you complete a challenge, you get to see other people's solutions to those challenges. Those solutions are then voted on and ranked for best practices and creativity. The UI for learning and building skills is epic in that.  

No money needs to trade hands for this because both sides benefit from improving their product and sharing their niche popularity. For example, more youth might get into code with Mindustry Logic and advance on codewars to professional coding skill proficiency while kids, mentors, and adults learning code on codewars can direct any one they want to encourage to code towards Mindustry as a way to eventually get into it.

**Before making this issue, place an `X` in the boxes below to confirm that you have acknowledged them.** *Failure to do so may result in your request being closed automatically.*

1. - [X] I have done a quick search in the list of suggestions to make sure this has not been suggested yet.
2. - [X] I have checked the [Trello](https://trello.com/b/aE2tcUwF/mindustry-trello) to make sure my suggestion isn't planned or implemented in a development version.
3. - [X] I am familiar with all the content already in the game or have glanced at the wiki to make sure my suggestion doesn't exist in the game yet.
4. - [X] I have read `README.md` to make sure my idea is not listed under the "A few things you shouldn't suggest" category.

**Comment by Bohaska** [comment]
Looks interesting and nice

**Comment by joshuaptfan** [comment]
As much as Mindustry Logic offers an opportunity to challenge young minds, I doubt that CodeWars would be interested in supporting a language that only exists in 1 game and has no real world applications.

**Comment by sample-text-here** [comment]
there are a few problems with this
- most blocks only have use in mindustry (eg. sensor, control, unit logic)
- unless codewars is willing to implement a graphic editor, you'd have to use text
- are there infinite memory cells? no memory cells? how would memory cells work?
- there are *much* better languages for both beginners and experts

overall, this is a poorly thought out

**Comment by Philosophist** [comment]
Re: Critics
1. Assembly language is what mLog is based on, therefore, the real world applications are vast. You just have to have the ability to imagine mLog as a gateway language to programming. 

2. "Better" languages to get people into are not fun languages for most gamers. Players who are first introduced to their own capability in programming by playing Mindustry are not playing Mindustry for the sake of learning. They might have people encouraging them to play for that reason, but the target audience I'm describing is motivated to learn when it's fun and not motivated to learn when it's work. 

3. The interface that I would suggest on the codewars website would be a small map portrayed from the game with limitations set by the challenge creator on what can be placed in the map if anything, the main focus should be the code challenge. Probably going to be one processor and some units to bind and/or buildings already connected and labelled. Memory cells would be limited by what the challenge creator places in the map or allows the user to place in the map. Hit run and see the tiny map window respond to the commands you wrote.

4. The whatever the in game visual editor happens to be when this feature is attempted that just be translated to the codewars website. There's also a chance that out-of-game visual editors with a more intuitive feel could become common tools for use among Mindustry players. It just depends on the growth of Mindustry by that time. 

5. Codewars is trying to gamify the learning process for code. That's their niche. Why would they be too good for a language that's primarily used for fun? People who want to learn code and are perfectly motivated on their own to do so don't tend to seek out codewars as having a particular edge in what they are looking for. They will take code challenges wherever they can find them.  People who struggle to learn because the prospects of learning code are overwhelming to them or because they need the leveling system to keep them motivated along the way, those are the people who get something special out of a challenge resource like codewars. 

6. If I didn't spend much time thinking about the suggestion, then I posted it thinking there's no harm in making the suggestion, and not thinking it out much was an intended feature of the post. I don't see it in that case as something to be ashamed of. 

If I did take a lot of time and energy to think out an idea, it's ok to be disagreed with on if it's a good idea, but the claim that it's "poorly thought out" comes from a place where the persona saying it doesn't know how much thought I put into it. It's a denial of the reality that I can observe where I obsessed with the thought and tried my best to work out the possible arguments against it before posting. I get that was just to say that you don't like the idea, but I did think this out and put some energy and respect into posting it. Maybe I'm just too much of a newb and you are leagues beyond me in your capability. I get that. Still, were that the case, I think the best practice would be to meet people where they are. Just a thought, and no [further] hard feelings about it.
